ledger/src/hardened_unpack.rs

```diff
@@ -0,0 +1,471 @@
+#[derive(Error, Debug)]
+pub enum UnpackError {
+    #[error("IO error")]
+    IO(#[from] std::io::Error),
+    #[error("Archive error")]
+    Archive(String),
+}
+
+pub type Result<T> = std::result::Result<T, UnpackError>;
+
+const MAX_SNAPSHOT_ARCHIVE_UNPACKED_SIZE: u64 = 500 * 1024 * 1024 * 1024; // 500 GiB
+const MAX_SNAPSHOT_ARCHIVE_UNPACKED_COUNT: u64 = 500_000;
+const MAX_GENESIS_ARCHIVE_UNPACKED_SIZE: u64 = 10 * 1024 * 1024; // 10 MiB
+const MAX_GENESIS_ARCHIVE_UNPACKED_COUNT: u64 = 100;
+
+fn checked_total_size_sum(total_size: u64, entry_size: u64, limit_size: u64) -> Result<u64> {
+    let total_size = total_size.saturating_add(entry_size);
+    if total_size > limit_size {
+        return Err(UnpackError::Archive(format!(
+            "too large snapshot: {:?}",
+            total_size
+        )));
+    }
+    Ok(total_size)
+}
+
+fn checked_total_count_increment(total_count: u64, limit_count: u64) -> Result<u64> {
+    let total_count = total_count + 1;
+    if total_count > limit_count {
+        return Err(UnpackError::Archive(format!(
+            "too many files in snapshot: {:?}",
+            total_count
+        )));
+    }
+    Ok(total_count)
+}
+
+fn check_unpack_result(unpack_result: bool, path: String) -> Result<()> {
+    if !unpack_result {
+        return Err(UnpackError::Archive(format!(
+            "failed to unpack: {:?}",
+            path
+        )));
+    }
+    Ok(())
+}
+
+fn unpack_archive<A: Read, P: AsRef<Path>, C>(
+    archive: &mut Archive<A>,
+    unpack_dir: P,
+    limit_size: u64,
+    limit_count: u64,
+    entry_checker: C,
+) -> Result<()>
+where
+    C: Fn(&[&str], tar::EntryType) -> bool,
+{
+    let mut total_size: u64 = 0;
+    let mut total_count: u64 = 0;
+
+    for entry in archive.entries()? {
+        let mut entry = entry?;
+        let path = entry.path()?;
+        let path_str = path.display().to_string();
+
+        // Although the `tar` crate safely skips at the actual unpacking, fail
+        // first by ourselves when there are odd paths like including `..` or /
+        // for our clearer pattern matching reasoning:
+        //   https://docs.rs/tar/0.4.26/src/tar/entry.rs.html#371
+        let parts = path.components().map(|p| match p {
+            CurDir => Some("."),
+            Normal(c) => c.to_str(),
+            _ => None, // Prefix (for Windows) and RootDir are forbidden
+        });
+        if parts.clone().any(|p| p.is_none()) {
+            return Err(UnpackError::Archive(format!(
+                "invalid path found: {:?}",
+                path_str
+            )));
+        }
+
+        let parts: Vec<_> = parts.map(|p| p.unwrap()).collect();
+        if !entry_checker(parts.as_slice(), entry.header().entry_type()) {
+            return Err(UnpackError::Archive(format!(
+                "extra entry found: {:?}",
+                path_str
+            )));
+        }
+        total_size = checked_total_size_sum(total_size, entry.header().size()?, limit_size)?;
```

Comment:
This intentionally uses **`Header::size()`** instead of `Header::entry_size()` following these documentation from the [reference](https://docs.rs/tar/0.4.26/tar/struct.Header.html#method.entry_size):

```
pub fn entry_size(&self) -> Result<u64>

Returns the size of entry's data this header represents.

This is different from Header::size for sparse files, which have some longer size() but shorter entry_size(). The entry_size() listed here should be the number of bytes in the archive this header describes.

May return an error if the field is corrupted.

```


[The used method's doc in this PR](https://docs.rs/tar/0.4.26/tar/struct.Header.html#method.size):

```
pub fn size(&self) -> Result<u64>

Returns the file size this header represents.

May return an error if the field is corrupted.
```

Also, as far as I read the source code, these documentation is consistent with the `tar` crate's implementation: https://docs.rs/tar/0.4.26/src/tar/header.rs.html#318